storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java

```diff
@@ -0,0 +1,390 @@
+        List<PreparableReporter> reporters = MetricsUtils.getPreparableReporters(conf);
+        for (PreparableReporter reporter : reporters) {
+            reporter.prepare(new MetricRegistry(), conf);
+            reporter.start();
+            LOG.info("Started statistics report plugin...");
+        }
+        if (handlerServer != null)
+            handlerServer.serve();
+    }
+
+    private void initClearThread() {
+        clearThread = Utils.asyncLoop(new Callable() {
+
+            @Override
+            public Object call() throws Exception {
+                for (Map.Entry<String, Integer> e : idtoStart.entrySet()) {
+
+                    if (Time.deltaSecs(e.getValue()) > Utils.getInt(conf.get(Config.DRPC_REQUEST_TIMEOUT_SECS), 0)) {
+                        String id = e.getKey();
+                        Semaphore sem = idtoSem.get(id);
+                        if (sem != null) {
+                            String func = idtoFunction.get(id);
+                            acquireQueue(func).remove(idtoRequest.get(id));
+                            LOG.warn("Timeout DRPC request id: {} start at {}", id, e.getValue());
+                            sem.release();
+                        }
+                        cleanup(id);
+                        LOG.info("Clear request " + id);
+                    }
+                }
+                return getTimeoutCheckSecs();
+            }
+        });
+    }
+
+    public Long getTimeoutCheckSecs() {
+        return timeoutCheckSecs;
+    }
+
+    public void launchServer(boolean isLocal, Map conf) throws Exception {
+
+        LOG.info("Starting drpc server for storm version {}", VersionInfo.getVersion());
+        this.conf = conf;
+        authorizer = mkAuthorizationHandler((String) (conf.get(Config.DRPC_AUTHORIZER)), conf);
+
+        initClearThread();
+        if (!isLocal)
+            initServer();
+    }
+
+    @Override
+    public void shutdown() {
+        meterShutdownCalls.mark();
+        clearThread.interrupt();
+    }
+
+    public void cleanup(String id) {
+        idtoSem.remove(id);
+        idtoResult.remove(id);
+        idtoStart.remove(id);
+        idtoFunction.remove(id);
+        idtoRequest.remove(id);
+    }
+
+    @Override
+    public String execute(String functionName, String funcArgs) throws DRPCExecutionException, AuthorizationException, org.apache.thrift.TException {
+        meterExecuteCalls.mark();
+        LOG.debug("Received DRPC request for {} {} at {} ", functionName, funcArgs, System.currentTimeMillis());
+        Map<String, String> map = new HashMap<>();
+        map.put(DRPCAuthorizerBase.FUNCTION_NAME, functionName);
+        checkAuthorization(authorizer, map, "execute");
+
+        int idinc = this.ctr.incrementAndGet();
+        int maxvalue = 1000000000;
+        int newid = idinc % maxvalue;
+        if (idinc != newid) {
+            this.ctr.compareAndSet(idinc, newid);
+        }
```

Comment:
This does not actually do the same thing.  what happens if the compareAndSet fails?  If you want it the same we have to loop around and try again until it succeeds.

```
int newid = 0;
do {
  int orig = ctr.get();
  newid = (orig + 1) % 1000000000;
} while (!ctr.compareAndSet(orig, newid));
String strid = String.valueOf(newid);
```